### PR TITLE
support tidb/tikv/pd configmap as arguments

### PR DESCRIPTION
**Pull request by Illyrix**
### What problem does this PR solve? 
Be able to choose a configmap for tidb/tikv/pd cluster
### Check List 

Tests 

 - Unit test
 - E2E test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has Go code change
 - Has CI related scripts change
 - Has Terraform scripts change

Side effects

 - Breaking backward compatibility

Related changes

 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
NONE
```

### CODE REVIEW

File: pkg/test-infra/tidb/operation.go [hunk 1]
```diff
@@ -439,20 +440,35 @@ func getTidbDiscoveryService(tc *v1alpha1.TidbCluster) *corev1.Service {
 	}
 }
 
+func readFileAsString(filename string) (string, error) {
+	if filename == "" {
+		return ``, nil
+	}
+	if bytes, err := ioutil.ReadFile(filename); err != nil {
+		return ``, err
+	} else {
```
[the remark this answers is not in the document]
Comment (yeya24) [reply pkg/test-infra/tidb/operation.go]: Using the patter is much clear here
```
bytes, err := ioutil.ReadFile(filename)
if err != nil {
  return "", err
}
return string(bytes), nil
```

File: pkg/test-infra/tidb/operation.go [hunk 2]
```diff
@@ -439,20 +440,35 @@ func getTidbDiscoveryService(tc *v1alpha1.TidbCluster) *corev1.Service {
...
+		return string(bytes), nil
+	}
+}
+
 // TODO: use the latest tidb-operator feature instead
```
Comment (yeya24) [inline pkg/test-infra/tidb/operation.go]: What does this TODO mean? Can we directly get the config from tidb-operator?

**Comment by yeya24** [comment]
I just found another problem.

``` Go
func (t *TidbOps) ApplyTiKVConfigMap(tc *v1alpha1.TidbCluster) error {
	configMap, err := getTiKVConfigMap(tc)
	if err != nil {
		return err
	}
	err = t.cli.Create(context.TODO(), configMap)
	if err != nil && !errors.IsAlreadyExists(err) {
		return err
	}
	return nil
}
```
Since the configmap we apply will not be cleaned up. So when the component loads the new config and try to create a new configmap, it fails because there is already one. So you should also update this function to use `controllerutil.CreateOrUpdate`. Example code:

``` Go
func (t *TidbOps) ApplyTiKVConfigMap(tc *v1alpha1.TidbCluster) error {
	configMap, err := getTiKVConfigMap(tc)
	if err != nil {
		return err
	}
	desired := configMap.DeepCopy()
	if _, err := controllerutil.CreateOrUpdate(context.TODO(), t.cli, configMap, func() error {
		configMap.Data = desired.Data
		return nil
	}); err != nil {
		return err
	}
	return nil
}
```

What do you think ? @Illyrix

**Comment by mahjonp** [comment]
> I just found another problem.
> 
> ```go
> func (t *TidbOps) ApplyTiKVConfigMap(tc *v1alpha1.TidbCluster) error {
> 	configMap, err := getTiKVConfigMap(tc)
> 	if err != nil {
> 		return err
> 	}
> 	err = t.cli.Create(context.TODO(), configMap)
> 	if err != nil && !errors.IsAlreadyExists(err) {
> 		return err
> 	}
> 	return nil
> }
> ```
> 
> Since the configmap we apply will not be cleaned up. So when the component loads the new config and try to create a new configmap, it fails because there is already one. So you should also update this function to use `controllerutil.CreateOrUpdate`. Example code:
> 
> ```go
> func (t *TidbOps) ApplyTiKVConfigMap(tc *v1alpha1.TidbCluster) error {
> 	configMap, err := getTiKVConfigMap(tc)
> 	if err != nil {
> 		return err
> 	}
> 	desired := configMap.DeepCopy()
> 	if _, err := controllerutil.CreateOrUpdate(context.TODO(), t.cli, configMap, func() error {
> 		configMap.Data = desired.Data
> 		return nil
> 	}); err != nil {
> 		return err
> 	}
> 	return nil
> }
> ```
> 
> What do you think? @Illyrix

So, @yeya24 does this on  https://github.com/pingcap/tipocket/pull/91, we can merge current PR before it.

**Pull request merged**